tests/modeladmin/tests.py

```diff
@@ -221,6 +221,29 @@ class BandAdmin(ModelAdmin):
             list(list(ma.get_formsets_with_inlines(request))[0][0]().forms[0].fields),
             ['main_band', 'opening_band', 'id', 'DELETE'])
 
+    def test_custom_formfield_override_readonly(self):
+        class AdminBandForm(forms.ModelForm):
+            name = forms.CharField()
+
+            class Meta:
+                exclude = tuple()
+                model = Band
+
+        class BandAdmin(ModelAdmin):
+            form = AdminBandForm
+            readonly_fields = ['name']
+
+        ma = BandAdmin(Band, self.site)
+
+        self.assertEqual(list(ma.get_form(request).base_fields),
+            ['bio', 'sign_date'])
+
+        self.assertEqual(list(ma.get_fields(request)),
+            ['bio', 'sign_date', 'name'])
+
+        self.assertEqual(list(ma.get_fieldsets(request)),
```

Comment:
It's used for the form rendering, although maybe it's not needed because `get_fieldset` implementation relies on `get_fields` in this version.